ENH: add ability to delete individual series, studies, patients

naucoin opened the pull request:
Add to the browser Remove button the information that it will
delete all selected series, studies, patients.

Add support for right click custom context menus in the
patient, study, series tables. Translate the local points
to global ones in terms of the table view and signal
right clicked with the global position so that the browser
can position the context menu appropriately.

Add the context menu to the DICOM browser to allow
deleting at just one level, with selection information.
Using the UIDs for now, but since they're not shown in
the tables, it can be changed.

Added a test for the DICOM browser, with the ability
to open it in interactive mode to test the right clicks.

Slicer bug:
http://www.na-mic.org/Bug/view.php?id=3792

When Slicer makes a custom DICOM browser, it moves the table views
directly into the new window, extracting them from the DICOM
table.
With this change, the table view takes care of mapping the point to
global, the table manager propagates the signal as a table specific
XRightClicked signal, and the browser responds to it with a table
specific context menu.

Slicer Issue #3792

pieper commented:
Looks good to me!

@nolden @AndreasFetzer do you want to have a look?

pieper commented:
Nicole and I just discussed this and we'll plan to merge it if we haven't heard an objection in a few days.

AndreasFetzer commented:
Hi Steve, sorry for the late response. He had a deadline on monday. I will have a look at it today.

AndreasFetzer commented:
The code looks good.

I have one remark:
In the context menu, the selected UIDs are displayed but are not displayed in the table views. Thus it is a useless information from an user's point of view since you cannot tell whether the right UID is selected.
I would suggest to display e.g. Patientname / Study or Series description instead of that in the context menu.

naucoin commented:
Thanks for the review! I'll update the informational string on the pop up menu to show your suggested fields.

fedorov commented:
@naucoin I tested with the current Slicer trunk and have this compile error. Did you see this? Probably not related to your change, but will need to be addressed anyway.

```
[ 75%] Building CXX object Modules/Loadable/Models/Widgets/CMakeFiles/qSlicerModelsModuleWidgets.dir/qMRMLModelDisplayNodeWidget.cxx.o
/Users/fedorov/github/Slicer/Modules/Loadable/Models/Widgets/qMRMLModelDisplayNodeWidget.cxx:82:31: error:
      no member named 'setNoneEnabled' in 'ctkVTKDataSetArrayComboBox'
  this->ActiveScalarComboBox->setNoneEnabled(true);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
1 error generated.
```

pieper commented:
Must be related to pull request #589

@lassoan what should Andrey do?

fedorov commented:
Ah, probably Nicole (or I) should rebase....

lassoan commented:
CTK was changed so you need to start a build at the top level (not just in the slicer-build subdirectory)

naucoin commented:
I've had to add some methods to the ctkDICOMDatabase in order to get at the descriptors, so I'm adding some extra testing as well. I should finish up on Monday.

naucoin commented:
@AndreasFetzer Let me know if the names and descriptions in the second commit are better.

fedorov commented:
@naucoin have you tested with a Slicer build, is it working there?

naucoin commented:
I had tested it in Slicer with a data set that had all the fields set, but in the database test I added it worked for one field empty, but I didn't update the browser test. Thanks for the report, I'll get it fixed ASAP.

naucoin commented:
I rebased and fixed the empty label bug here:
https://github.com/naucoin/CTK/tree/3792-unable-to-remove-one-series-of-an-exam-rebase

fedorov commented:
Ok, this one works fine with Slicer, thanks!

It seems to work great! 

Few suggestion I have is to remove the list of selected items from the popup, and remove "Cancel" option. The first one looks more like debugging, and can be overwhelming when the number of selected rows is large. The "Cancel" option is unusual to see in a popup menu. One click outside to cancel is equal to clicking cancel option, so the benefit is unclear.

![image](https://cloud.githubusercontent.com/assets/313942/9505176/e3c845e0-4c0e-11e5-92a8-b4a6809af7f5.png)

fedorov commented:
If you want to communicate to the user the fact that several items will be deleted, maybe it is better to just say "Delete <#number> selected <series|studies|patients>".

naucoin commented:
The pop up covers some of the selected rows, so I wanted to give the user some feedback about what was going to be deleted.
Good point about the Cancel, I'll remove it.

fedorov commented:
> The pop up covers some of the selected rows, so I wanted to give the user some feedback about what was going to be deleted.

I still think it is not a good idea to include those names in the popup. Also note that if you remove these "Selected" pieces, the popup will cover much less.

jcfr commented:
+1 for "Delete <#number> selected series".

If you would like, you could still show a confirmation popup listing the
series with the option of not showing it the following time ...

On Wed, Aug 26, 2015 at 4:44 PM, Andrey Fedorov <notifications@github.com>
wrote:

> The pop up covers some of the selected rows, so I wanted to give the user
> some feedback about what was going to be deleted.
>
> I still think it is not a good idea to include those names in the popup.
> Also note that if you remove these "Selected" pieces, the popup will cover
> much less.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-135165497>.
>



-- 
+1 919 869 8849

fedorov commented:
> If you would like, you could still show a confirmation popup listing the
series with the option of not showing it the following time ...

+1

naucoin commented:
I'm personally fine if the consensus is to just show the number of items to be deleted, I'm going to be out of the office tomorrow through Monday so any new iterations may end up waiting until next week.

naucoin commented:
I'll have to revisit this next week, I've encountered a bug where the mouse clicks are behaving oddly (takes two left clicks to select, the right click only works on the patients table).

jcfr commented:
Sounds good. Have good travel and talk to you next week.
Jc

On Wed, Aug 26, 2015 at 5:58 PM, Nicole Aucoin <notifications@github.com>
wrote:

> I'll have to revisit this next week, I've encountered a bug where the
> mouse clicks are behaving oddly (takes two left clicks to select, the right
> click only works on the patients table).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-135182797>.
>



-- 
+1 919 869 8849

fedorov commented:
Ditto!

---
Sent from my mobile device
On Aug 26, 2015 6:02 PM, "Jean-Christophe Fillion-Robin" <
notifications@github.com> wrote:

> Sounds good. Have good travel and talk to you next week.
> Jc
>
> On Wed, Aug 26, 2015 at 5:58 PM, Nicole Aucoin <notifications@github.com>
> wrote:
>
> > I'll have to revisit this next week, I've encountered a bug where the
> > mouse clicks are behaving oddly (takes two left clicks to select, the
> right
> > click only works on the patients table).
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/commontk/CTK/pull/587#issuecomment-135182797>.
> >
>
>
>
> --
> +1 919 869 8849
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-135184742>.
>

fedorov commented:
@naucoin can you also add the capability to export selected stuff? So the idea would be to add "Export to file system" option to popup, and when clicked, show Directory open dialog, and when selected, just copy the files from their current location to the user specified location. While doing this, we can also create a user-readable patient/study/series hierarchy of the directories for the exported stuff.

I thought this is a very similar feature to be added in the same PR, or feel free to create a new PR.

pieper commented:
+1 on this feature.

For reference on the structure of the output directory you can look at this
dicomsort code [1].  It has a very configurable method for defining the
naming, but in practice 99% of the time people will be fine with the
default structure so I'd suggest just starting with this format or
something similar:

targetDir +
'/%PatientName/%StudyDesciption-%StudyDate/%SeriesDescription-%SeriesNumber-%InstanceNumber.dcm'




[1] https://github.com/pieper/dicomsort

On Mon, Aug 31, 2015 at 9:03 AM, Andrey Fedorov <notifications@github.com>
wrote:

> @naucoin <https://github.com/naucoin> can you also add the capability to
> export selected stuff? So the idea would be to add "Export to file system"
> option to popup, and when clicked, show Directory open dialog, and when
> selected, just copy the files from their current location to the user
> specified location. While doing this, we can also create a user-readable
> patient/study/series hierarchy of the directories for the exported stuff.
>
> I thought this is a very similar feature to be added in the same PR, or
> feel free to create a new PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-136364820>.
>

fedorov commented:
@naucoin if you follow @pieper advice on the directory structure, it probably makes sense to convert to ASCII and remove special characters first.

pieper commented:
Agreed with @fedorov here's a routine you can inspire from:

https://github.com/pieper/dicomsort/blob/master/dicomsort.py#L99-L108

On Mon, Aug 31, 2015 at 10:32 AM, Andrey Fedorov <notifications@github.com>
wrote:

> @naucoin <https://github.com/naucoin> if you follow @pieper
> <https://github.com/pieper> advice on the directory structure, it
> probably makes sense to convert to ASCII and remove special characters
> first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-136389217>.
>

jcfr commented:
@naucoin I would suggest you integrate your current work and create an other PR.

@pieper Thanks for pointing out that useful snippet.

This made me think about the following ... 

What are the best practice regarding DICOM name clean up ? ... I anticipate copy/pasting the [code pointed](https://github.com/pieper/dicomsort/blob/master/dicomsort.py#L99-L108) by @pieper to cause us trouble moving forward (aka maintenance pain + bug hard to find). What do you think if we add a function to pydicom, DCMTK or CTK or in some other a common location ... ?

pieper commented:
@jcfr - agreed, it is a tricky thing to make useful filenames from dicom
elements.  The heuristic in the code I pointed to is the best I could come
up with at the time, and of course it is very much OS platform dependent
(even dependent on the specific filesystem where you are trying to write)
and I didn't find a perfect guideline to follow.  So I would argue the
right place to add this logic would be in Qt and/or in python's os module
and then make use of it from DICOM.  As a practical matter probably CTK
Core is a logical place for us to put the logic.

On Mon, Aug 31, 2015 at 11:54 AM, Jean-Christophe Fillion-Robin <
notifications@github.com> wrote:

> @naucoin <https://github.com/naucoin> I would suggest you integrate your
> current work and create an other PR.
>
> @pieper <https://github.com/pieper> Thanks for pointing out that useful
> snippet.
>
> This made me think about the following ...
>
> What are the best practice regarding DICOM name clean up ? ... I
> anticipate copy/pasting the code pointed
> <https://github.com/pieper/dicomsort/blob/master/dicomsort.py#L99-L108>
> by @pieper <https://github.com/pieper> to cause us trouble moving forward
> (aka maintenance pain + bug hard to find). What do you think if we add a
> function to pydicom, DCMTK or CTK or in some other a common location ... ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/commontk/CTK/pull/587#issuecomment-136413693>.
>

fedorov commented:
Just a note to Nicole, please make sure you augment that code with the conversion to ASCII. Things like superscripts are not uncommon in study descriptions.

naucoin commented:
The export will definitely be a separate issue and pull request, please create a new issue or bug and assign it to me.
Taking another look at the right click issues to get this pull request finished up.

fedorov commented:
Here is the issue: http://www.na-mic.org/Bug/view.php?id=3163. I know it is assigned to Csaba, and he did a lot of work on this in the Subject hierarchy, but I also think Isaiah was looking for something much simpler, without any edits to the DICOM, just context menu and export to directory.

lassoan commented:
There are several data repositories (Slicer scene in memory, Slicer scene on disk, DICOM database, DICOM files not indexed by DICOM database) and altogether 12 possible save/load/import/export operations between them. Csaba implemented these two "DICOM export" operations:
* export from "Slicer scene in memory" to DICOM files not indexed by DICOM database
* save from "Slicer scene in memory" to DICOM database

fedorov commented:
Andras, the functionality suggested here is the standard DICOM export that is available in all PACS workstations: locate files in the database, and copy to the user-specified location. I agree there are many options. I think this feature will fill an important gap.

lassoan commented:
Yes, I completely agree. I just wanted to clarify that the scope of Csaba's work was different.

naucoin commented:
I re-rebased this branch, fixed the right click bug and added a confirmation dialogue with the patient name, study or series description:
https://github.com/naucoin/CTK/tree/3792-unable-to-remove-one-series-of-an-exam-rebase
I also updated the browser test to avoid the copy/link on import pop up.
Any feedback on that branch? I'll want to squash it and make a new pull request, closing this one.

fedorov commented:
@naucoin thanks! I will run rebuild overnight and plan to test tomorrow.

fedorov commented:
Build finished quicker than I thought.

Looks great to me in Slicer and is ready to merge AFAIK! Nice work! :+1:

jcfr commented on Libs/DICOM/Widgets/ctkDICOMBrowser.cpp:
```diff
@@ -657,3 +667,168 @@ void ctkDICOMBrowser::onModelSelected(const QItemSelection &item1, const QItemSe
+      {
+      selectionLabel = new QLabel(seriesDescription);
+      }
+    if (!selectionLabel->text().isEmpty())
```
nitpick: Could be simpler to simply set a variable `QString selectionDescription;`

That code would then become:

```
if (!selectionDescription.isEmpty())
  {
  QWidgetAction *selectionAction = new QWidgetAction(menu);
  selectionAction->setDefaultWidget(new QLabel(selectionDescription));
  menu->addAction(selectionAction);
  }
```

naucoin replied:
This has been redone in the -rebase branch I linked in my last comment, please take a look here:
https://github.com/commontk/CTK/commit/e13f457c0316bba96213a8daf53c522d9c08f884

jcfr commented:
@naucoin To facilitate review, I created a new PR based on the new branch `3792-unable-to-remove-one-series-of-an-exam-rebase`. I am closing this one.

jcfr commented:
See https://github.com/commontk/CTK/pull/597

naucoin commented:
I had wanted to squash it down to one commit before making a pull request, will you still want me to do that before it gets integrated?
